### PR TITLE
Add support for non-float32 normalization for linen normalization layers

### PR DESCRIPTION
**Pull request by jheek**
Previously we always casted to float32 even if
this is a downcast. For example f64 -> f32 or
complex -> float. This change maxes the
reduction dtype *at least* float32

### CODE REVIEW

File: flax/linen/normalization.py [hunk 1]
```diff
@@ -37,6 +37,13 @@ def _canonicalize_axes(rank: int, axes: Axes) -> Iterable[int]:
     axes = (axes,)
   return tuple(set([rank + axis if axis < 0 else axis for axis in axes]))
 
+def _norm_sq(x):
+  """Computes the elementwise square norm |x|^2."""
```
Comment (marcvanzee) [inline flax/linen/normalization.py:41]: This just seems to compute `x^2`? I don't see the norm being computed anywhere, or am I missing something?

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Are you sure your PR description is correct? I can see this PR changes the behavior for complex numbers, but for f64 it seems to be a no op.

Also, you changed `x^2` to `|x|^2` (the square of the norm?), but the code just seems to use `lax.square`, which just computes the square element-wise.

File: flax/linen/normalization.py (hunk 1 above)
Comment (jheek) [reply flax/linen/normalization.py:41]: Typically you use norm/abs value to disambiguate between x^2 = (a+bi)(a+bi) vs x^2=(a+bi)(a-bi) = a^2 + b^2. For real numbers it both amounts to normal squaring indeed.

File: flax/linen/normalization.py [hunk 2]
```diff
@@ -46,25 +53,27 @@ def _compute_stats(x: Array, axes: Axes,
...
+    by using Var = E[|x|^2] - |E[x]|^2 instead of Var = E[|x - E[x]|^2]).
   - Clips negative variances to zero which can happen due to
     roundoff errors. This avoids downstream NaNs.
   - Supports averaging across a parallel axis and subgroups of a parallel axis
     with a single `lax.pmean` call to avoid latency.
   """
-  x = jnp.asarray(x, jnp.float32)
+  # promote x to at least float32, this avoids half precision computation
+  # but preserves double or complex floating points
+  x = jnp.asarray(x, jnp.promote_types(jnp.float32, jnp.result_type(x)))
```
Comment (jheek) [inline flax/linen/normalization.py:64]: @marcvanzee  this is the part that affects both complex and float64. We cast x to float32 before computing the statistics. I changed that to promoting to at least float32 so we don't downcast f64 and complex numbers

File: flax/linen/normalization.py (hunk 1 above)
Comment (PhilipVinc) [reply flax/linen/normalization.py:41]: This function is usually called `abs2` in C++ Eigen, Julia, and has recently been added to SciPy as `abs_sq`, so maybe that would be a better name

**Review by PhilipVinc — APPROVED** [review]
Thanks for this! Other than the nitpick about naming of `_norm_sq` this looks good to me.

File: flax/linen/normalization.py (hunk 1 above)
Comment (jheek) [reply flax/linen/normalization.py]: I agree

**Review by marcvanzee — APPROVED** [review]
